Source/WebGPU/WebGPU/ComputePipeline.mm

```diff
@@ -86,14 +86,37 @@ static MTLSize metalSize(auto workgroupSize)
     if (!function)
         return ComputePipeline::createInvalid(*this);
 
-    MTLComputePipelineReflection *reflection;
-    bool hasBindGroups = pipelineLayout.numberOfBindGroupLayouts() > 0;
-    auto computePipelineState = createComputePipelineState(m_device, function, pipelineLayout, computeInformation, label, &reflection, hasBindGroups ? MTLPipelineOptionNone : MTLPipelineOptionArgumentInfo);
+    if (!pipelineLayout.numberOfBindGroupLayouts() && entryPointInformation.defaultLayout) {
```

Comment:
I don't understand why 0 bind group layouts is being used in an `if` statement. The number of bind group layouts, be it 0 or nonzero, should never be compared for equality. If an author deletes their last BGL in their code, that shouldn't cause the possibility of going down a different codepath.

I also don't think `entryPointInformation.defaultLayout` should be in the branch condition either. `entryPointInformation` is information returned by the compiler to us, but the source of truth about whether or not we should be using the default layout is passed in from the API. If the API tells us that we should be using the default layout, then _that's_ when we should be interrogating `entryPointInformation.defaultLayout` from the compiler. (Ideally, there would also be some way to tell the compiler "don't even bother generating default layout information, if the API isn't telling us that we need it.)

Reply:
0 bind groups layouts is being used in the if statement to represent `auto` layout.

`auto` layout has 0 bind groups, it tells us we need to consult the layout of the vertex and fragment shaders (or just compute for compute pipeline) when auto layout is passed.

Consider this call to `GPUDevice.createComputePipeline`:
```
device.createComputePipeline({
      layout: 'auto',
      compute: {
        module: computeShader,
        entryPoint: 'main',
      },
    });
```

`pipelineLayout` corresponds to what is passed to `layout`. But if `layout` == `auto`, we have zero layout information, so we must generate one from the compute shader module itself.

Reply:
I guess I mean: What if the shader legitimately has 0 bind groups? That's totally valid, and doesn't mean we should be using auto layout.

Reply:
I was unaware of this, I will file a bugzilla to fix that up.

Reply:
`PipelineLayout::isValid` returns false if there are zero layouts:
```
    // FIXME: Is it impossible to legitimately have no bind group layouts in a valid object?
    bool isValid() const { return !m_bindGroupLayouts.isEmpty(); }
```

but I can not find anywhere in the specification that says a pipeline layout must have at least one bind group layout.

Reply:
Filed https://bugs.webkit.org/show_bug.cgi?id=257978 to address GPUAutoLayout mode